dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,626 @@
+
+    payload_location = get_blob_from_archive(layers[0]["digest"], tmppath, archive)
+    with open(payload_location, "r") as f:
+        payload = json.load(f)
+        image_name = payload["critical"]["identity"]["docker-reference"]
+
+    return image_name, signatures
+
+
+def get_remote_log_index(image_str: str) -> int:
+    """Check the registry to find the remote log index for the given image"""
+    try:
+        signature_manifest = registry.get_signature_manifest(image_str)
+        return get_log_index_from_oci_signatures(signature_manifest)
+    except:
+        return 0
+
+
+def get_log_index_from_oci_signatures(signature_manifest: dict) -> int:
+    """
+    Extracts the log index from OCI signatures without having
+    to parse and verify the whole signature.
+    """
+    remote_log_index = 0
+    for layer in signature_manifest["layers"]:
+        bundle = json.loads(layer["annotations"]["dev.sigstore.cosign/bundle"])
+        log_index = bundle["Payload"]["logIndex"]
+        if log_index > remote_log_index:
+            remote_log_index = log_index
+    return remote_log_index
+
+
+def get_file_digest(
+    path: Optional[Path] = None, content: Optional[bytes] = None
+) -> str:
+    """Get the sha256 digest of a file or content"""
+    if not path and not content:
+        raise errors.UpdaterError("No file or content provided")
+    if path:
+        with path.open("rb") as f:
+            content = f.read()
+    if content:
+        return sha256(content).hexdigest()
+    return ""
+
+
+def load_and_verify_signatures(
+    image_digest: str,
+    pubkey: Path,
+    bypass_verification: bool = False,
+    signatures_path: Optional[Path] = None,
+) -> List[Dict]:
+    """
+    Load signatures from the local filesystem
+
+    See store_signatures() for the expected format.
+    """
+    if not signatures_path:
+        signatures_path = SIGNATURES_PATH
+
+    pubkey_signatures = signatures_path / get_file_digest(pubkey)
+    if not pubkey_signatures.exists():
+        msg = (
+            f"Cannot find a '{pubkey_signatures}' folder. "
+            "You might need to download the image signatures first."
+        )
+        raise errors.SignaturesFolderDoesNotExist(msg)
+
+    signatures_file = pubkey_signatures / f"{image_digest}.json"
+
+    if not signatures_file.exists():
+        msg = (
+            f"Cannot find a '{signatures_file}' file. "
+            "You might need to download the image signatures first."
+        )
+        raise errors.LocalSignatureNotFound(msg)
+
+    with open(signatures_file) as f:
+        log.debug("Loading signatures from %s", f.name)
+        signatures = json.load(f)
+
+    if not bypass_verification:
+        verify_signatures(signatures, image_digest, pubkey)
+
+    return signatures
+
+
+def store_signatures(
+    signatures: list[Dict],
+    image_digest: str,
+    pubkey: Path = DEFAULT_PUBKEY_LOCATION,
+    update_logindex: bool = True,
+) -> None:
+    """
```

Comment:
We might want to actually do the verification of the signatures here rather than requiring them to be verified before calling this without having a way to verify it's the case.

One of the concerns is probably the fact that we don't want to verify the signatures too many times, so ensure it's not he case.